Firestore/tests/System/AggregateQueryTest.php

```diff
@@ -0,0 +1,223 @@
+
+        $q = $collection->endBefore(end($refs)->snapshot());
+        $this->assertQueryCount($q, count($refs)-1);
+
+        $q = $collection->endAt(end($refs)->snapshot());
+        $this->assertQueryCount($q, count($refs));
+    }
+
+    public function testLimitToLast()
+    {
+        // limit is not supported by emulator
+        $this->skipEmulatorTests();
+        $collection = self::$client->collection(uniqid(self::COLLECTION_NAME));
+        self::$localDeletionQueue->add($collection);
+        for ($i = 1; $i <= 5; $i++) {
+            $collection->add(['i' => $i]);
+        }
+
+        $q = $collection->orderBy('i')
+            ->limitToLast(2);
+
+        $this->assertQueryCount($q, 2);
+    }
+
+    public function testLimitToLastWithCursors()
+    {
+        // cursors are not supported by emulator
+        $this->skipEmulatorTests();
+        $collection = self::$client->collection(uniqid(self::COLLECTION_NAME));
+        self::$localDeletionQueue->add($collection);
+        for ($i = 1; $i <= 5; $i++) {
+            $collection->add(['i' => $i]);
+        }
+
+        $q = $collection->orderBy('i')
+            ->startAt([2])
+            ->endAt([4])
+            ->limitToLast(5);
+
+        $this->assertQueryCount($q, 3);
+    }
+
+    private function insertDoc(array $fields)
+    {
+        return $this->query->add($fields);
+    }
+
+    private function assertQueryCount(Query $query, $expectedCount)
+    {
+        $actualCount = $query->count();
+        $this->assertEquals($expectedCount, $actualCount);
+
+        $this->assertQueryCountWithMultipleAggregations($query, $expectedCount);
+    }
+
+    private function assertQueryCountWithMultipleAggregations(
+        Query $query,
+        $expectedCount
+    ) {
+        $aggregations = [
+            Aggregate::count()->alias('count'),
+            Aggregate::count()->alias('count_upto_1')->limit(1),
+            Aggregate::count()->alias('count_upto_2')->limit(2),
+        ];
+        $expectedCounts = [
+            'count' => $expectedCount,
+            'count_upto_1' => min($expectedCount, 1),
+            'count_upto_2' => min($expectedCount, 2),
+        ];
+        foreach ($aggregations as $aggregation) {
+            $query = $query->addAggregation($aggregation);
+        }
+
+        $snapshot = $query->getSnapshot();
+
+        foreach ($expectedCounts as $k => $v) {
+            $expectedCount = $v;
+            $actualCount = $snapshot->get($k);
+            $this->assertEquals($expectedCount, $actualCount);
+        }
+
+        $this->assertEquals(0, strlen($snapshot->getTransaction()));
+        $actualTimestamp = $snapshot->getReadTime();
+        $expectedTimestamp = new Timestamp(new \DateTimeImmutable());
+        $this->assertInstanceOf(Timestamp::class, $snapshot->getReadTime());
+        $this->assertLessThanOrEqual(
+            $actualTimestamp->get()->getTimestamp(),
+            $expectedTimestamp->get()->getTimestamp()
+        );
```

Comment:
this needs to change to something like assert with delta, because emulator gives latest time, rather than copy-pasting readtime param from inputs to output.

Reply:
Seems this is fixed in latest emulator version!